### ISSUE TITLE
SemiSpace with side forwarding bits should clear side metadata before each GC.

### ISSUE THREAD

**Issue by wenyuzhao** [mmtk/mmtk-core#339]
Currently, CopySpace assumes that the forwarding bits for all objects are zero at the start of each GC. This is true for in-header forwarding bits, but not for side forwarding bits -- they may be preserved to the next GC if not zeroed properly.

The potential fixes are:

Either clear the forwarding bits after each object allocation, at `CopySpace:: initialize_object_metadata`,

or bulk zeroing the whole side forwarding table at `CopySpace::prepare`.

**Comment by qinsoon** [comment]
I think we should do bulk zeroing. But we do not have side metadata spec for fowarding bits in `master`. Maybe at the time when we add side metadata specs for forwarding bits, we should consider this. I don't see this is a bug right now, and I am not clear what we can do about it.

**Comment by qinsoon** [comment]
Does this bug show up when we turn off GC byte in header? @wenyuzhao

**Comment by wenyuzhao** [comment]
> Does this bug show up when we turn off GC byte in header? @wenyuzhao

Yes. Currently I turn off header GC bytes for V8 and see this problem. I made a quick bulk zeroing fix to make it work with side gc byte.

The fix works for me at the moment, but as a proper fix, I haven't yet figured out how to do bulk zeroing for a space with FreeListPageResource. Although CopySpace only uses MonotonePageResource, but probably future GCs that relies on FreeListPageResource also need such pre-zeroing as well. 

Since we are rewriting the gc-byte interface, I think this is better to be fixed within the new interface.

**Comment by qinsoon** [comment]
OK. I see.

@javadamiri You probably want to consider this in your gc-byte refactoring.

**Comment by wks** [comment]
This was fixed in https://github.com/mmtk/mmtk-core/pull/373

Closing...

**Comment by wks** [comment]
> ... The fix works for me at the moment, but as a proper fix, I haven't yet figured out how to do bulk zeroing for a space with FreeListPageResource. ...

Since https://github.com/mmtk/mmtk-core/pull/974/files, the ImmixSpace zeroes on-the-side forwarding bits for each block, in the `PrepareBlockState` work packet.
